### PR TITLE
block migration for unsupported configurations

### PR DESCRIPTION
**Pull request by ozdanborne**
## Description

- tolerations
- volume mounts
- canal

I also converted typha into a pointer in the components struct since it is an optional component.

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/controller/migration/convert/aws-cni-policy-only_test.go [hunk 1]
```diff
@@ -192,6 +192,16 @@ func awsCNIPolicyOnlyConfig() []runtime.Object {
 			Spec: appsv1.DeploymentSpec{
 				Template: corev1.PodTemplateSpec{
 					Spec: corev1.PodSpec{
+						Tolerations: []corev1.Toleration{
```
Comment (tmjd) [inline pkg/controller/migration/convert/aws-cni-policy-only_test.go:195]: I messed up here, actually this Deployment needs to be removed, there is no calico-kube-controllers in the amazon-vpc-cni repo.
This file was meant to represent https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/master/config/master/calico.yaml. I think I had copied this and forgot to remove it.

File: pkg/controller/migration/convert/core.go [hunk 2]
```diff
@@ -145,3 +216,21 @@ func handleFelixNodeMetrics(c *components, install *Installation) error {
...
+func checkTolerations(existing []corev1.Toleration, expected ...corev1.Toleration) error {
+	if len(existing) != len(expected) {
+		return fmt.Errorf("missing expected tolerations. have: %+v. expecting: %+v", existing, expected)
+	}
+
+OUTER:
+	for _, t := range expected {
+		for _, k := range existing {
+			if k == t {
+				continue OUTER
```
Comment (tmjd) [inline pkg/controller/migration/convert/core.go:229]: This is just to continue the outer loop so doing a break here should work.

Comment (ozdanborne) [reply pkg/controller/migration/convert/core.go:229]: I don't think that's correct - If I break the inner loop, I'll immediately hit the `return fmt.Errorf`

File: pkg/controller/migration/convert/aws-cni-policy-only_test.go (hunk 1 above)
Comment (ozdanborne) [reply pkg/controller/migration/convert/aws-cni-policy-only_test.go]: I added a commit which removes the kubeController deployment from this test, makes the kubeController in components a pointer, and doesn't throw an error in `getComponents` if it doesn't detect the kubecontrollers deployment :+1:

File: pkg/controller/migration/convert/core.go [hunk 3]
```diff
@@ -71,6 +78,76 @@ func handleCore(c *components, install *Installation) error {
 		install.Spec.FlexVolumePath = "None"
 	}
 
+	// Volumes for lib-modules, xtables-lock, var-run-calico, var-lib-calico, or policysync have been changed
+	v := getVolume(c.node.Spec.Template.Spec, "lib-modules")
+	if v == nil || v.HostPath == nil || v.HostPath.Path != "/lib/modules" {
+		return ErrIncompatibleCluster{"must mount /lib/modules from host"}
```
Comment (tmjd) [inline pkg/controller/migration/convert/core.go:84]: Should this message be more like `expected calico-node to have Volume 'lib-modules' with host path '/lib/modules'`? (same for the others below too)

File: pkg/controller/migration/convert/core.go [hunk 4]
```diff
@@ -71,6 +78,76 @@ func handleCore(c *components, install *Installation) error {
...
+		},
+		corev1.Toleration{
+			Effect:   corev1.TaintEffectNoSchedule,
+			Operator: corev1.TolerationOpExists,
+		},
+		corev1.Toleration{
+			Effect:   corev1.TaintEffectNoExecute,
+			Operator: corev1.TolerationOpExists,
+		}); err != nil {
+		return fmt.Errorf("calico-node has incompatible tolerations: %v", err)
```
Comment (tmjd) [inline pkg/controller/migration/convert/core.go:123]: should be ErrIncompatibleCluster

File: pkg/controller/migration/convert/core.go [hunk 5]
```diff
@@ -71,6 +78,76 @@ func handleCore(c *components, install *Installation) error {
...
+		if err := checkTolerations(c.kubeControllers.Spec.Template.Spec.Tolerations,
+			corev1.Toleration{
+				Key:      "CriticalAddonsOnly",
+				Operator: corev1.TolerationOpExists,
+			},
+			corev1.Toleration{
+				Effect: corev1.TaintEffectNoSchedule,
+				Key:    "node-role.kubernetes.io/master",
+			}); err != nil {
+			return fmt.Errorf("kube-controllers has incompatible tolerations: %v", err)
```
Comment (tmjd) [inline pkg/controller/migration/convert/core.go:137]: should be ErrIncompatibleCluster

File: pkg/controller/migration/convert/core.go [hunk 6]
```diff
@@ -71,6 +78,76 @@ func handleCore(c *components, install *Installation) error {
...
+		}
+	}
+
+	// check typha tolerations
+	if c.typha != nil {
+		if err := checkTolerations(c.typha.Spec.Template.Spec.Tolerations, corev1.Toleration{
+			Key:      "CriticalAddonsOnly",
+			Operator: corev1.TolerationOpExists,
+		}); err != nil {
+			return fmt.Errorf("typha has incompatible tolerations: %v", err)
```
Comment (tmjd) [inline pkg/controller/migration/convert/core.go:147]: ditto

File: pkg/controller/migration/convert/core.go (hunk 2 above)
Comment (tmjd) [reply pkg/controller/migration/convert/core.go:235]: Yep you're right.
Do you like this pattern? It is just a GOTO which I always feel should be avoided if possible.
One alternative would be to use a bool for if the toleration is found and use that to skip the `return fmt.Errorf`, then break would work.
But if you're not bothered by the 'goto' I can live with it.

File: pkg/controller/migration/convert/core.go [hunk 7]
```diff
@@ -145,3 +222,21 @@ func handleFelixNodeMetrics(c *components, install *Installation) error {
 
 	return nil
 }
+
+func checkTolerations(existing []corev1.Toleration, expected ...corev1.Toleration) error {
+	if len(existing) != len(expected) {
+		return fmt.Errorf("missing expected tolerations. have: %+v. expecting: %+v", existing, expected)
```
Comment (tmjd) [inline pkg/controller/migration/convert/core.go:228]: The errors in this function should be ErrIncompatibleCluster.

**Pull request merged**